diff --git a/gce/deploy.sh b/gce/deploy.sh
new file mode 100644
index 0000000..61b2846
--- /dev/null
+++ b/gce/deploy.sh
@@ -0,0 +1,151 @@
+# Copyright 2015 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#! /bin/bash
+
+set -ex
+
+ZONE=us-central1-f
+
+GROUP=frontend-group
+TEMPLATE=$GROUP-tmpl
+MACHINE_TYPE=f1-micro
+STARTUP_SCRIPT=my-startup.sh
+IMAGE=ubuntu-15-04
+SCOPES="userinfo-email,\
+logging-write,\
+storage-full,\
+datastore,\
+https://www.googleapis.com/auth/projecthosting"
+TAGS=http-server
+
+MIN_INSTANCES=1
+MAX_INSTANCES=10
+TARGET_UTILIZATION=0.6
+
+SERVICE=frontend-web-service
+
+#
+# Instance group setup
+#
+
+# First we have to create an instance template.
+# This template will be used by the instance group
+# to create new instances.
+
+# [START create_template]
+gcloud compute instance-templates create $TEMPLATE \
+  --machine-type $MACHINE_TYPE \
+  --scopes $SCOPES \
+  --metadata-from-file startup-script=$STARTUP_SCRIPT \
+  --image $IMAGE \
+  --tags $TAGS
+# [END create_template]
+
+# Create the managed instance group.
+
+# [START create_group]
+gcloud compute instance-groups managed \
+  create $GROUP \
+  --base-instance-name $GROUP \
+  --size $MIN_INSTANCES \
+  --template $TEMPLATE \
+  --zone $ZONE
+# [END create_group]
+
+# [START create_named_port]
+gcloud compute instance-groups managed set-named-ports \
+    $GROUP \
+    --named-port http:8080 \
+    --zone $ZONE
+# [END create_named_port]
+
+#
+# Load Balancer Setup
+#
+
+# A complete HTTP load balancer is structured as follows:
+#
+# 1) A global forwarding rule directs incoming requests to a target HTTP proxy.
+# 2) The target HTTP proxy checks each request against a URL map to determine the
+#    appropriate backend service for the request.
+# 3) The backend service directs each request to an appropriate backend based on
+#    serving capacity, zone, and instance health of its attached backends. The
+#    health of each backend instance is verified using either a health check.
+#
+# We'll create these resources in reverse order:
+# service, health check, backend service, url map, proxy.
+
+# Create a health check
+# The load balancer will use this check to keep track of which instances to send traffic to.
+# Note that health checks will not cause the load balancer to shutdown any instances.
+
+# [START create_health_check]
+gcloud compute http-health-checks create ah-health-check \
+  --request-path /_ah/health
+# [END create_health_check]
+
+# Create a backend service, associate it with the health check and instance group.
+# The backend service serves as a target for load balancing.
+
+# [START create_backend_service]
+gcloud compute backend-services create $SERVICE \
+  --http-health-check ah-health-check
+# [END create_backend-service]
+
+# [START add_backend_service]
+gcloud compute backend-services add-backend $SERVICE \
+  --group $GROUP \
+  --zone $ZONE
+# [END add_backend_service]
+
+# Create a URL map and web Proxy. The URL map will send all requests to the
+# backend service defined above.
+
+# [START create_url_map]
+gcloud compute url-maps create $SERVICE-map \
+  --default-service $SERVICE
+# [END create_url_map]
+
+# [START create_http_proxy]
+gcloud compute target-http-proxies create $SERVICE-proxy \
+  --url-map $SERVICE-map
+# [END create_http_proxy]
+
+# Create a global forwarding rule to send all traffic to our proxy
+
+# [START create_forwarding_rule]
+gcloud compute forwarding-rules create $SERVICE-http-rule \
+  --global \
+  --target-http-proxy $SERVICE-proxy \
+  --port-range 80
+# [END create_forwarding_rule]
+
+#
+# Autoscaler configuration
+#
+# [START set_autoscaling]
+gcloud compute instance-groups managed set-autoscaling \
+  $GROUP \
+  --max-num-replicas $MAX_INSTANCES \
+  --target-load-balancing-utilization $TARGET_UTILIZATION \
+  --zone $ZONE
+# [END set_autoscaling]
+
+# [START create_firewall]
+gcloud compute firewall-rules create default-allow-http-8080 \
+    --allow tcp:8080 \
+    --source-ranges 0.0.0.0/0 \
+    --target-tags http-server \
+    --description "Allow port 8080 access to http-server"
+# [END create_firewall]
diff --git a/gce/nginx/bookshelf.conf b/gce/nginx/bookshelf.conf
new file mode 100644
index 0000000..ccd912b
--- /dev/null
+++ b/gce/nginx/bookshelf.conf
@@ -0,0 +1,49 @@
+# Copyright 2015 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+server {
+    listen       80;
+    root         /opt/app/bookshelf/web;
+    index index.html index.htm index.php;
+
+    # Media: images, icons, video, audio, HTC
+    location ~* \.(?:jpg|jpeg|gif|png|ico|cur|gz|svg|svgz|mp4|ogg|ogv|webm|htc)$ {
+      expires 10m;
+      access_log off;
+      etag on;
+      add_header Cache-Control "public";
+    }
+
+    # CSS and Javascript
+    location ~* \.(?:css|js)$ {
+      expires 10m;
+      access_log off;
+      etag on;
+      add_header Cache-Control "public";
+    }
+
+    # pass the PHP scripts to FastCGI server
+    location ~ \.php$ {
+      try_files $uri =404;
+      fastcgi_split_path_info ^(.+?\.php)(/.*)$;
+      fastcgi_pass unix:/var/run/php5-fpm.sock;
+      fastcgi_index index.php;
+      include fastcgi_params;
+    }
+
+    location / {
+      # try to serve file directly, fallback to front controller
+      try_files $uri /index.php$is_args$args;
+    }
+}
diff --git a/gce/nginx/fastcgi_params b/gce/nginx/fastcgi_params
new file mode 100644
index 0000000..417ff18
--- /dev/null
+++ b/gce/nginx/fastcgi_params
@@ -0,0 +1,41 @@
+# Copyright 2015 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+fastcgi_param   QUERY_STRING            $query_string;
+fastcgi_param   REQUEST_METHOD          $request_method;
+fastcgi_param   CONTENT_TYPE            $content_type;
+fastcgi_param   CONTENT_LENGTH          $content_length;
+
+fastcgi_param   SCRIPT_FILENAME         $document_root$fastcgi_script_name;
+fastcgi_param   SCRIPT_NAME             $fastcgi_script_name;
+fastcgi_param   PATH_INFO               $fastcgi_path_info;
+fastcgi_param   PATH_TRANSLATED         $document_root$fastcgi_path_info;
+fastcgi_param   REQUEST_URI             $request_uri;
+fastcgi_param   DOCUMENT_URI            $document_uri;
+fastcgi_param   DOCUMENT_ROOT           $document_root;
+fastcgi_param   SERVER_PROTOCOL         $server_protocol;
+
+fastcgi_param   GATEWAY_INTERFACE       CGI/1.1;
+fastcgi_param   SERVER_SOFTWARE         nginx/$nginx_version;
+
+fastcgi_param   REMOTE_ADDR             $http_x_real_ip;
+fastcgi_param   REMOTE_PORT             $remote_port;
+fastcgi_param   SERVER_ADDR             $server_addr;
+fastcgi_param   SERVER_PORT             $server_port;
+fastcgi_param   SERVER_NAME             $server_name;
+
+fastcgi_param   HTTPS                   $https;
+
+# PHP only, required if PHP was built with --enable-force-cgi-redirect
+fastcgi_param   REDIRECT_STATUS         200;
diff --git a/gce/startup-script.sh b/gce/startup-script.sh
new file mode 100644
index 0000000..385aa80
--- /dev/null
+++ b/gce/startup-script.sh
@@ -0,0 +1,90 @@
+#!/bin/bash
+
+# Copyright 2015 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# [START all]
+set -e
+
+export HOME=/root
+
+# Talk to the metadata server to get the project id
+PROJECTID=$(curl -s "http://metadata.google.internal/computeMetadata/v1/project/project-id" -H "Metadata-Flavor: Google")
+
+# [START logging]
+curl -s "https://storage.googleapis.com/signals-agents/logging/google-fluentd-install.sh" | bash
+cat >/etc/google-fluentd/config.d/bookshelf.conf << EOF
+<source>
+  type tail
+  format none
+  path /opt/app/log/*.log
+  pos_file /var/tmp/fluentd.app.pos
+  read_from_head true
+  rotate_wait 10s
+  tag bookshelf
+</source>
+EOF
+service google-fluentd restart &
+# [END logging]
+
+# [START php]
+# Install PHP and dependencies from apt
+apt-get update
+apt-get install -y git nginx mongodb-clients php5 php5-fpm php5-dev php-pear pkg-config
+pecl install mongodb
+
+# enable the pecl-installed mongodb extension
+cat >/etc/php5/mods-available/20-mongodb.ini << EOF
+extension=mongodb.so
+EOF
+
+ln -s /etc/php5/mods-available/20-mongodb.ini /etc/php5/fpm/conf.d/20-mongodb.ini
+ln -s /etc/php5/mods-available/20-mongodb.ini /etc/php5/cli/conf.d/20-mongodb.ini
+
+# Install composer
+curl -sS https://getcomposer.org/installer | \
+    /usr/bin/php -- \
+    --install-dir=/usr/local/bin \
+    --filename=composer
+
+# Get the source code
+git config --global credential.helper gcloud.sh
+
+# Change branch from master if not using master
+git clone https://source.developers.google.com/p/$PROJECTID /opt/app -b master
+
+# run composer
+composer install -d /opt/app/bookshelf --no-ansi --no-progress
+# [END php]
+
+# [START project-config]
+# pull our custom config file from the metadata server and add it to the project
+curl -s "http://metadata.google.internal/computeMetadata/v1/instance/attributes/project-config" \
+  -H "Metadata-Flavor: Google" >> /opt/app/bookshelf/config/settings.yml
+# [END project-config]
+
+# [START nginx]
+# disable the default nginx configuration
+rm /etc/nginx/sites-enabled/default
+
+# enable our nginx configuration
+cp /opt/app/bookshelf/gce/nginx/bookshelf.conf /etc/nginx/sites-available/bookshelf.conf
+ln -s /etc/nginx/sites-available/bookshelf.conf /etc/nginx/sites-enabled/bookshelf.conf
+
+# add fastcgi params for PHP
+cp /opt/app/bookshelf/gce/nginx/fastcgi_params /etc/nginx/fastcgi_params
+
+systemctl restart nginx.service
+# [END nginx]
+# [END all]
diff --git a/gce/teardown.sh b/gce/teardown.sh
new file mode 100644
index 0000000..7452db1
--- /dev/null
+++ b/gce/teardown.sh
@@ -0,0 +1,41 @@
+# Copyright 2015 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#! /bin/bash
+
+set -x
+
+ZONE=us-central1-f
+gcloud config set compute/zone $ZONE
+
+GROUP=frontend-group
+TEMPLATE=$GROUP-tmpl
+SERVICE=frontend-web-service
+
+gcloud compute instance-groups managed stop-autoscaling $GROUP --zone $ZONE
+
+gcloud compute forwarding-rules delete $SERVICE-http-rule --global 
+
+gcloud compute target-http-proxies delete $SERVICE-proxy 
+
+gcloud compute url-maps delete $SERVICE-map 
+
+gcloud compute backend-services delete $SERVICE 
+
+gcloud compute http-health-checks delete ah-health-check
+
+gcloud compute instance-groups managed delete $GROUP  
+
+gcloud compute instance-templates delete $TEMPLATE 
+
+gcloud compute firewall-rules delete default-allow-http-8080
