diff --git a/.dockerignore b/.dockerignore
deleted file mode 100644
index 305612c..0000000
--- a/.dockerignore
+++ /dev/null
@@ -1,2 +0,0 @@
-test/*
-coverage/*
diff --git a/Dockerfile b/Dockerfile
deleted file mode 100644
index 4af0b3c..0000000
--- a/Dockerfile
+++ /dev/null
@@ -1,12 +0,0 @@
-# The Google App Engine php runtime is Debian Jessie with PHP installed
-# and various os-level packages to allow installation of popular PHP
-# libraries. The source is on github at:
-#   https://github.com/GoogleCloudPlatform/php-docker
-FROM gcr.io/google_appengine/php
-
-# Add our NGINX and php.ini config
-ENV DOCUMENT_ROOT=${APP_DIR}/web
-
-# Workaround for AUFS-related permission issue:
-# See https://github.com/docker/docker/issues/783#issuecomment-56013588
-RUN cp -R ${APP_DIR} ${APP_DIR}-copy; rm -r ${APP_DIR}; mv ${APP_DIR}-copy ${APP_DIR}; chmod -R 550 ${APP_DIR}; chown -R root.www-data ${APP_DIR}
diff --git a/Dockerfile.worker b/Dockerfile.worker
deleted file mode 100644
index 6e596f8..0000000
--- a/Dockerfile.worker
+++ /dev/null
@@ -1,8 +0,0 @@
-# The Google App Engine php runtime is Debian Jessie with PHP installed
-# and various os-level packages to allow installation of popular PHP
-# libraries. The source is on github at:
-#   https://github.com/GoogleCloudPlatform/php-docker
-FROM gcr.io/google_appengine/php
-
-# override the default CMD for the pubsub worker
-CMD php bin/pubsub/entrypoint.php
diff --git a/README.md b/README.md
index 760f1b3..9d89dda 100644
--- a/README.md
+++ b/README.md
@@ -1,8 +1,8 @@
-# Deploying to Google Kubernetes Engine
+# Deploying to Google Compute Engine
 
-This folder contains the sample code for the [Deploying to Google Kubernetes Engine][tutorial-gke]
+This folder contains the sample code for the [Deploying to Google Compute Engine][tutorial-gce]
 tutorial. Please refer to the tutorial for instructions on configuring, running,
 and deploying this sample.
 
-[tutorial-gke]: https://cloud.google.com/php/tutorials/bookshelf-on-container-engine
+[tutorial-gce]: https://cloud.google.com/php/getting-started/run-on-compute-engine
 
diff --git a/bookshelf-frontend.yaml b/bookshelf-frontend.yaml
deleted file mode 100644
index 041b17c..0000000
--- a/bookshelf-frontend.yaml
+++ /dev/null
@@ -1,33 +0,0 @@
-# This file configures the bookshelf application frontend. The frontend serves
-# public web traffic.
-
-apiVersion: extensions/v1beta1
-kind: Deployment
-metadata:
-  name: bookshelf-frontend
-  labels:
-    app: bookshelf
-# The bookshelf frontend replica set ensures that at least 3
-# instances of the bookshelf app are running on the cluster.
-# For more info about Pods see:
-#   https://cloud.google.com/container-engine/docs/pods/
-spec:
-  replicas: 3
-  template:
-    metadata:
-      labels:
-        app: bookshelf
-        tier: frontend
-    spec:
-      containers:
-      - name: bookshelf-app
-        # Replace $GCLOUD_PROJECT with your project ID or use `make template`.
-        image: gcr.io/$GCLOUD_PROJECT/bookshelf
-        # This setting makes nodes pull the docker image every time before
-        # starting the pod. This is useful when debugging, but should be turned
-        # off in production.
-        imagePullPolicy: Always
-        # The bookshelf process listens on port 8080 for web traffic by default.
-        ports:
-        - name: http-server
-          containerPort: 8080
diff --git a/bookshelf-service.yaml b/bookshelf-service.yaml
deleted file mode 100644
index 45290ce..0000000
--- a/bookshelf-service.yaml
+++ /dev/null
@@ -1,23 +0,0 @@
-
-# The bookshelf service provides a load-balancing proxy over the bookshelf
-# frontend pods. By specifying the type as a 'LoadBalancer', Container Engine
-# will create an external HTTP load balancer.
-# For more information about Services see:
-#   https://cloud.google.com/container-engine/docs/services/
-# For more information about external HTTP load balancing see:
-#   https://cloud.google.com/container-engine/docs/load-balancer
-apiVersion: v1
-kind: Service
-metadata:
-  name: bookshelf-frontend
-  labels:
-    app: bookshelf
-    tier: frontend
-spec:
-  type: LoadBalancer
-  ports:
-  - port: 80
-    targetPort: http-server
-  selector:
-    app: bookshelf
-    tier: frontend
diff --git a/bookshelf-worker.yaml b/bookshelf-worker.yaml
deleted file mode 100644
index a5c0af3..0000000
--- a/bookshelf-worker.yaml
+++ /dev/null
@@ -1,47 +0,0 @@
-# Copyright 2015 Google Inc.
-#
-# Licensed under the Apache License, Version 2.0 (the "License");
-# you may not use this file except in compliance with the License.
-# You may obtain a copy of the License at
-#
-#     http://www.apache.org/licenses/LICENSE-2.0
-#
-# Unless required by applicable law or agreed to in writing, software
-# distributed under the License is distributed on an "AS IS" BASIS,
-# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
-# See the License for the specific language governing permissions and
-# limitations under the License
-
-# This file configures the bookshelf task worker. The worker is responsible
-# for processing book requests and updating book information.
-
-apiVersion: extensions/v1beta1
-kind: Deployment
-metadata:
-  name: bookshelf-worker
-  labels:
-    app: bookshelf
-# The bookshelf worker replica set ensures that at least 2 instances of the
-# bookshelf worker pod are running on the cluster.
-# For more info about Pods see:
-#   https://cloud.google.com/container-engine/docs/pods/
-spec:
-  replicas: 2
-  template:
-    metadata:
-      labels:
-        app: bookshelf
-        tier: worker
-    spec:
-      containers:
-      - name: bookshelf-app
-        # Replace $GCLOUD_PROJECT with your project ID or use `make template`.
-        image: gcr.io/$GCLOUD_PROJECT/bookshelf-worker
-        # This setting makes nodes pull the docker image every time before
-        # starting the pod. This is useful when debugging, but should be turned
-        # off in production.
-        imagePullPolicy: Always
-        # The bookshelf process listens on port 8080 for web traffic by default.
-        ports:
-        - name: ratchet-server
-          containerPort: 8080
diff --git a/gce/deploy.sh b/gce/deploy.sh
new file mode 100644
index 0000000..b1b8102
--- /dev/null
+++ b/gce/deploy.sh
@@ -0,0 +1,151 @@
+# Copyright 2015 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#! /bin/bash
+
+set -ex
+
+ZONE=us-central1-f
+
+GROUP=frontend-group
+TEMPLATE=$GROUP-tmpl
+MACHINE_TYPE=f1-micro
+STARTUP_SCRIPT=my-startup.sh
+IMAGE=ubuntu-15-04
+SCOPES="userinfo-email,\
+logging-write,\
+storage-full,\
+datastore,\
+https://www.googleapis.com/auth/projecthosting"
+TAGS=http-server
+
+MIN_INSTANCES=1
+MAX_INSTANCES=10
+TARGET_UTILIZATION=0.6
+
+SERVICE=frontend-web-service
+
+#
+# Instance group setup
+#
+
+# First we have to create an instance template.
+# This template will be used by the instance group
+# to create new instances.
+
+# [START create_template]
+gcloud compute instance-templates create $TEMPLATE \
+  --machine-type $MACHINE_TYPE \
+  --scopes $SCOPES \
+  --metadata-from-file startup-script=$STARTUP_SCRIPT \
+  --image $IMAGE \
+  --tags $TAGS
+# [END create_template]
+
+# Create the managed instance group.
+
+# [START create_group]
+gcloud compute instance-groups managed \
+  create $GROUP \
+  --base-instance-name $GROUP \
+  --size $MIN_INSTANCES \
+  --template $TEMPLATE \
+  --zone $ZONE
+# [END create_group]
+
+# [START create_named_port]
+gcloud compute instance-groups managed set-named-ports \
+    $GROUP \
+    --named-ports http:8080 \
+    --zone $ZONE
+# [END create_named_port]
+
+#
+# Load Balancer Setup
+#
+
+# A complete HTTP load balancer is structured as follows:
+#
+# 1) A global forwarding rule directs incoming requests to a target HTTP proxy.
+# 2) The target HTTP proxy checks each request against a URL map to determine the
+#    appropriate backend service for the request.
+# 3) The backend service directs each request to an appropriate backend based on
+#    serving capacity, zone, and instance health of its attached backends. The
+#    health of each backend instance is verified using either a health check.
+#
+# We'll create these resources in reverse order:
+# service, health check, backend service, url map, proxy.
+
+# Create a health check
+# The load balancer will use this check to keep track of which instances to send traffic to.
+# Note that health checks will not cause the load balancer to shutdown any instances.
+
+# [START create_health_check]
+gcloud compute http-health-checks create ah-health-check \
+  --request-path /_ah/health
+# [END create_health_check]
+
+# Create a backend service, associate it with the health check and instance group.
+# The backend service serves as a target for load balancing.
+
+# [START create_backend_service]
+gcloud compute backend-services create $SERVICE \
+  --http-health-checks ah-health-check
+# [END create_backend-service]
+
+# [START add_backend_service]
+gcloud compute backend-services add-backend $SERVICE \
+  --group $GROUP \
+  --zone $ZONE
+# [END add_backend_service]
+
+# Create a URL map and web Proxy. The URL map will send all requests to the
+# backend service defined above.
+
+# [START create_url_map]
+gcloud compute url-maps create $SERVICE-map \
+  --default-service $SERVICE
+# [END create_url_map]
+
+# [START create_http_proxy]
+gcloud compute target-http-proxies create $SERVICE-proxy \
+  --url-map $SERVICE-map
+# [END create_http_proxy]
+
+# Create a global forwarding rule to send all traffic to our proxy
+
+# [START create_forwarding_rule]
+gcloud compute forwarding-rules create $SERVICE-http-rule \
+  --global \
+  --target-http-proxy $SERVICE-proxy \
+  --port-range 80
+# [END create_forwarding_rule]
+
+#
+# Autoscaler configuration
+#
+# [START set_autoscaling]
+gcloud compute instance-groups managed set-autoscaling \
+  $GROUP \
+  --max-num-replicas $MAX_INSTANCES \
+  --target-load-balancing-utilization $TARGET_UTILIZATION \
+  --zone $ZONE
+# [END set_autoscaling]
+
+# [START create_firewall]
+gcloud compute firewall-rules create default-allow-http-8080 \
+    --allow tcp:8080 \
+    --source-ranges 0.0.0.0/0 \
+    --target-tags http-server \
+    --description "Allow port 8080 access to http-server"
+# [END create_firewall]
diff --git a/gce/fluentd/bookshelf.conf b/gce/fluentd/bookshelf.conf
new file mode 100644
index 0000000..98bd80e
--- /dev/null
+++ b/gce/fluentd/bookshelf.conf
@@ -0,0 +1,10 @@
+<source>
+  type tail
+  format none
+  path /opt/app/log/*.log
+  pos_file /var/tmp/fluentd.app.pos
+  read_from_head true
+  rotate_wait 10s
+  tag bookshelf
+</source>
+
diff --git a/gce/nginx/bookshelf.conf b/gce/nginx/bookshelf.conf
new file mode 100644
index 0000000..a54a9de
--- /dev/null
+++ b/gce/nginx/bookshelf.conf
@@ -0,0 +1,49 @@
+# Copyright 2015 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+server {
+    listen       80;
+    root         /opt/app/web;
+    index index.html index.htm index.php;
+
+    # Media: images, icons, video, audio, HTC
+    location ~* \.(?:jpg|jpeg|gif|png|ico|cur|gz|svg|svgz|mp4|ogg|ogv|webm|htc)$ {
+      expires 10m;
+      access_log off;
+      etag on;
+      add_header Cache-Control "public";
+    }
+
+    # CSS and Javascript
+    location ~* \.(?:css|js)$ {
+      expires 10m;
+      access_log off;
+      etag on;
+      add_header Cache-Control "public";
+    }
+
+    # pass the PHP scripts to FastCGI server
+    location ~ \.php$ {
+      try_files $uri =404;
+      fastcgi_split_path_info ^(.+?\.php)(/.*)$;
+      fastcgi_pass unix:/var/run/php5-fpm.sock;
+      fastcgi_index index.php;
+      include fastcgi_params;
+    }
+
+    location / {
+      # try to serve file directly, fallback to front controller
+      try_files $uri /index.php$is_args$args;
+    }
+}
diff --git a/gce/nginx/fastcgi_params b/gce/nginx/fastcgi_params
new file mode 100644
index 0000000..417ff18
--- /dev/null
+++ b/gce/nginx/fastcgi_params
@@ -0,0 +1,41 @@
+# Copyright 2015 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+fastcgi_param   QUERY_STRING            $query_string;
+fastcgi_param   REQUEST_METHOD          $request_method;
+fastcgi_param   CONTENT_TYPE            $content_type;
+fastcgi_param   CONTENT_LENGTH          $content_length;
+
+fastcgi_param   SCRIPT_FILENAME         $document_root$fastcgi_script_name;
+fastcgi_param   SCRIPT_NAME             $fastcgi_script_name;
+fastcgi_param   PATH_INFO               $fastcgi_path_info;
+fastcgi_param   PATH_TRANSLATED         $document_root$fastcgi_path_info;
+fastcgi_param   REQUEST_URI             $request_uri;
+fastcgi_param   DOCUMENT_URI            $document_uri;
+fastcgi_param   DOCUMENT_ROOT           $document_root;
+fastcgi_param   SERVER_PROTOCOL         $server_protocol;
+
+fastcgi_param   GATEWAY_INTERFACE       CGI/1.1;
+fastcgi_param   SERVER_SOFTWARE         nginx/$nginx_version;
+
+fastcgi_param   REMOTE_ADDR             $http_x_real_ip;
+fastcgi_param   REMOTE_PORT             $remote_port;
+fastcgi_param   SERVER_ADDR             $server_addr;
+fastcgi_param   SERVER_PORT             $server_port;
+fastcgi_param   SERVER_NAME             $server_name;
+
+fastcgi_param   HTTPS                   $https;
+
+# PHP only, required if PHP was built with --enable-force-cgi-redirect
+fastcgi_param   REDIRECT_STATUS         200;
diff --git a/gce/startup-script.sh b/gce/startup-script.sh
new file mode 100644
index 0000000..bf6298c
--- /dev/null
+++ b/gce/startup-script.sh
@@ -0,0 +1,78 @@
+#!/bin/bash
+
+# Copyright 2015 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# [START all]
+set -e
+export HOME=/root
+
+# [START php]
+# Install PHP and dependencies from apt
+apt-get update
+apt-get install -y git nginx mongodb-clients php5 php5-fpm php5-mysql php5-dev php-pear pkg-config
+pecl install mongodb
+
+# Enable the MongoDB PHP extension
+echo "extension=mongodb.so" >> /etc/php5/mods-available/mongodb.ini
+php5enmod mongodb
+
+# Install Composer
+curl -sS https://getcomposer.org/installer | \
+    /usr/bin/php -- \
+    --install-dir=/usr/local/bin \
+    --filename=composer
+
+# Fetch the project ID from the Metadata server
+PROJECTID=$(curl -s "http://metadata.google.internal/computeMetadata/v1/project/project-id" -H "Metadata-Flavor: Google")
+
+# Get the application source code
+git config --global credential.helper gcloud.sh
+git clone https://source.developers.google.com/p/$PROJECTID /opt/src -b master
+ln -s /opt/src/optional-compute-engine /opt/app
+
+# Run Composer
+composer install -d /opt/app --no-ansi --no-progress
+# [END php]
+
+# [START project_config]
+# Fetch the application config file from the Metadata server and add it to the project
+curl -s "http://metadata.google.internal/computeMetadata/v1/instance/attributes/project-config" \
+  -H "Metadata-Flavor: Google" >> /opt/app/config/settings.yml
+# [END project_config]
+
+# [START nginx]
+# Disable the default NGINX configuration
+rm /etc/nginx/sites-enabled/default
+
+# Enable our NGINX configuration
+cp /opt/app/gce/nginx/bookshelf.conf /etc/nginx/sites-available/bookshelf.conf
+ln -s /etc/nginx/sites-available/bookshelf.conf /etc/nginx/sites-enabled/bookshelf.conf
+cp /opt/app/gce/nginx/fastcgi_params /etc/nginx/fastcgi_params
+
+# Start NGINX
+systemctl restart nginx.service
+# [END nginx]
+
+# [START logging]
+# Install Fluentd
+curl -s "https://storage.googleapis.com/signals-agents/logging/google-fluentd-install.sh" | bash
+
+# Enable our Fluentd configuration
+cp /opt/app/gce/fluentd/bookshelf.conf /etc/google-fluentd/config.d/bookshelf.conf
+
+# Start Fluentd
+service google-fluentd restart &
+# [END logging]
+# [END all]
diff --git a/gce/teardown.sh b/gce/teardown.sh
new file mode 100644
index 0000000..7452db1
--- /dev/null
+++ b/gce/teardown.sh
@@ -0,0 +1,41 @@
+# Copyright 2015 Google Inc.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#! /bin/bash
+
+set -x
+
+ZONE=us-central1-f
+gcloud config set compute/zone $ZONE
+
+GROUP=frontend-group
+TEMPLATE=$GROUP-tmpl
+SERVICE=frontend-web-service
+
+gcloud compute instance-groups managed stop-autoscaling $GROUP --zone $ZONE
+
+gcloud compute forwarding-rules delete $SERVICE-http-rule --global 
+
+gcloud compute target-http-proxies delete $SERVICE-proxy 
+
+gcloud compute url-maps delete $SERVICE-map 
+
+gcloud compute backend-services delete $SERVICE 
+
+gcloud compute http-health-checks delete ah-health-check
+
+gcloud compute instance-groups managed delete $GROUP  
+
+gcloud compute instance-templates delete $TEMPLATE 
+
+gcloud compute firewall-rules delete default-allow-http-8080
diff --git a/test/app-e2e.yaml b/test/app-e2e.yaml
index 4725e7b..8dffa4d 100644
--- a/test/app-e2e.yaml
+++ b/test/app-e2e.yaml
@@ -1,4 +1,4 @@
-runtime: custom
+runtime: php
 env: flex
 
 runtime_config:
diff --git a/test/composer-mongodb.json b/test/composer-mongodb.json
index fd2d873..bf8e118 100644
--- a/test/composer-mongodb.json
+++ b/test/composer-mongodb.json
@@ -7,7 +7,6 @@
   "require": {
     "php": "~5.5.0|~5.6.0|~7.0.0",
     "google/apiclient": "~2.0@RC",
-    "google/cloud": "^0.2",
     "monolog/monolog": "^1.17",
     "mongodb/mongodb": "1.0.4",
     "silex/silex": "^1.3",
diff --git a/test/lib/E2eTest.php b/test/lib/E2eTest.php
index b4b5925..6578ab5 100644
--- a/test/lib/E2eTest.php
+++ b/test/lib/E2eTest.php
@@ -24,7 +24,7 @@ abstract class E2eTest extends \PHPUnit_Framework_TestCase
 {
     public static function setUpBeforeClass()
     {
-        // TODO: deploy to GKE
+        // TODO: deploy to GCE
     }
 
     public static function tearDownAfterClass()
